### ISSUE TITLE
Allow to limit the scope of the EventBus

### ISSUE THREAD

**Issue by netopyr** [canoo/dolphin-platform#137]
Currently every event is sent to all clients. We would like to send events to subscribers in the same session only. (UseCase: a global NotificationController receives messages to show toasts in the application.)

We could implement this ourselves creating ad-hoc topics for each client, but it would be convenient if the platform offers out-of-the-box functionality to limit the reach. This could be done either by offering different event buses or an additional parameter.

**Comment by netopyr** [comment]
Depends on #127 and #50

**Comment by hendrikebbers** [comment]
I think we need 3 different types: Global Scope, Session Scope, Client Scope.
I ask myself if there are scenarios where you want to send a message to all other session without the current one?

**Comment by netopyr** [comment]
We have implemented a small solution ourselves now and as it turns out, we need several entry points to the DolphinEventBus which are scoped differently. We have one entry point, which is a @Service and can send only global events to all clients. And we have a second entry point, which is a @Component and can send global as well as session-only events and allows to subscribe to global and session events.

The latter actually needs to be @Client-scoped, but we were not able to implement that so far.

It may be useful, if the global entry point could also subscribe to global events, but currently this is not possible, because the DolphinEventBus does not allow subscriptions from outside of a session.

tl;dr
I think we need three components to access the DolphinEventBus: a @Service-scoped one, that can send and receive global events, a @Session-scoped one that can send and receive global as well as session events and a @Client-scoped one that can send and receive all kinds of events.

**Comment by hendrikebbers** [comment]
Mmmh, I think that we need the functionality that you describe but instead of your solution I would prefer a event bus that is defined as client scoped and provide methods to send messages to specific scopes (global, session, client)

**Comment by netopyr** [comment]
But if it is client-scoped, how do use it in a bean that is @Service-scoped?

**Comment by hendrikebbers** [comment]
Ah, yes. Spring works in a different way than JavaEE/CDI here. In CDI it's no problem to inject an instance with a smaller scope (for example inject request scope object in servlet scope object) since all injected instances are wrapped by a proxy. We need to check if we can do this in Spring by adding a specific implementation at the DP sources (like "use proxy for this object"). It's important that a developer do not need to do something specific here. What do you think?

**Comment by netopyr** [comment]
It depends on the details. Usually I prefer to be forced to think for a while about the scope and the functionality that is available, if the alternative is that everything blows up no earlier than during runtime.

If you have specific adapters for each scope, you can make it more obvious what kind of events you can publish. For instance a service-scoped adapter could have only one method publishGlobalEvent() while a session-scoped adapter could have an additional method publishSessionEvent().

But overall I do not have a strong opinion here as long as I can quickly find the reason, if something is wrong.

**Comment by hendrikebbers** [comment]
From my point of view it will be much harder for a developer to work with several event bus classes instead of having a central class that provides the difference types of distribution as methods.

**Comment by hendrikebbers** [comment]
Oh, my first answer was directly after my holiday without thinking about it 😆 
I think the single event bus class should be defined as application scoped.
